mmcv/ops/csrc/pytorch/cpu/rotated_feature_align.cpp

```diff
@@ -0,0 +1,275 @@
+// modified from
+// https://github.com/SJTU-Thinklab-Det/r3det-on-mmdetection/blob/master/mmdet/ops/fr/src/feature_refine_kernel.cu
+#include "pytorch_cpp_helper.hpp"
+#include "pytorch_device_registry.hpp"
+
+template <typename T>
+T bilinear_interpolate(const T* input, const int height, const int width, T y,
+                       T x, const int index /* index for debug only*/) {
+  // deal with cases that inverse elements are out of feature map boundary
+  if (y < -1.0 || y > height || x < -1.0 || x > width) return 0;
+
+  if (y <= 0) y = 0;
+  if (x <= 0) x = 0;
+
+  int y_low = (int)y;
+  int x_low = (int)x;
+  int y_high;
+  int x_high;
+
+  if (y_low >= height - 1) {
+    y_high = y_low = height - 1;
+    y = (T)y_low;
+  } else {
+    y_high = y_low + 1;
+  }
+
+  if (x_low >= width - 1) {
+    x_high = x_low = width - 1;
+    x = (T)x_low;
+  } else {
+    x_high = x_low + 1;
+  }
+
+  T ly = y - y_low;
+  T lx = x - x_low;
+  T hy = 1. - ly, hx = 1. - lx;
+  // do bilinear interpolation
+  T v1 = input[y_low * width + x_low];
+  T v2 = input[y_low * width + x_high];
+  T v3 = input[y_high * width + x_low];
+  T v4 = input[y_high * width + x_high];
+  T w1 = hy * hx, w2 = hy * lx, w3 = ly * hx, w4 = ly * lx;
```

Comment:
Advice: interpolate can be done with [fma](https://en.cppreference.com/w/cpp/numeric/math/fma). Some times it would be faster.

Reply:
Done

Reply:
Errr, Nope, that is not what I mean. Doing fma on int won't be a good idea. [this](https://developer.nvidia.com/blog/lerp-faster-cuda/) is an example.

You can try this:
```c++
  const T v_low = fma(v2 - v1, lx, v1);
  const T v_high = fma(v4 - v3, lx, v3);
  const T val = fma(v_high - v_low, ly, v_low);
```